### PR TITLE
Implement client protocol v2

### PR DESCRIPTION
**Pull request by AlekseiS**
Please see the following wiki page for v2 protocol description and key differences compared to v1: https://github.com/prestodb/presto/wiki/Presto-v2-client-protocol

Limitations of the current implementation:
* Client v2 only supports one download channel.
* Information about desired parallelism needs to be passed from a client. Parallelism of the last stage will be used.
* Information about desired parallelism needs to be taken into account by planner to add an arbitrary buffer on top. Currently, it works without additional buffers.
* Information about desired parallelism needs to be taken into account by scheduler to schedule the right number of arbitrary buffer tasks.
* Information about using coordinator only download needs to be passed from the client and used by planner/scheduler to fulfil it.
* Certain DDL query shapes need to output "true" as a result of their operation.
* jdbc needs to be able to select v2 version. Currently, it can only be done in CLI.
* Json representation of structural types is the same as in v1. There were some thoughs about representing it differently.
* "Actions", such as "set transaction id", "clear session properties", etc are represented the same way as in v1.
* Query is started immediately when submitted. There's a consideration to start a query on the first status request.
* Need to clarify/change how data uris are being sent to a client. Currently, they will be sent on every update in a cumulative way, but only once each time.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java [hunk 1]
```diff
@@ -60,184 +51,99 @@
...
-        return result;
+        QueryInfo queryInfo = queryManager.createQuery(sessionContext, query);
+        ExchangeClient exchangeClient = exchangeClientSupplier.get(new SimpleLocalMemoryContext(new AggregatedMemoryContext()));
+        Query createdQuery = new QueryV1(queryInfo, queryManager, sessionPropertyManager, exchangeClient, dataProcessorExecutor, timeoutExecutor, blockEncodingSerde);
+        createdQuery.queryManager.addOutputInfoListener(createdQuery.queryId, createdQuery::setQueryOutputInfo);
+        return createdQuery;
     }
 
-    private Query(
+    public static Query createV2(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java]: Why not to extract `QueryFactory`? Then it would be nice to have an interface of  `Query` defined and if needed `AbstractQuery` as well.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/StatementResourceV2.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static io.airlift.http.server.AsyncResponseHandler.bindAsyncResponse;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+@Path("/v2/statement")
+public class StatementResourceV2
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/StatementResourceV2.java]: I do not see `StatementResouceV1` defined. It would be nice to follow the same convention as for `Query`.

File: presto-client/src/main/java/com/facebook/presto/client/StatementClientV2.java [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;
+import static java.net.HttpURLConnection.HTTP_UNAVAILABLE;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+@ThreadSafe
+public class StatementClientV2
```
Comment (kokosing) [inline presto-client/src/main/java/com/facebook/presto/client/StatementClientV2.java]: ah... here it comes...

File: presto-client/src/main/java/com/facebook/presto/client/StatementClientFactory.java [hunk 4]
```diff
@@ -15,12 +15,29 @@
 
 import okhttp3.OkHttpClient;
 
+import static java.util.Objects.requireNonNull;
+
 public final class StatementClientFactory
 {
+    public static final ProtocolVersion DEFAULT_PROTOCOL_VERSION = ProtocolVersion.V1;
```
Comment (kokosing) [inline presto-client/src/main/java/com/facebook/presto/client/StatementClientFactory.java]: I would remove this static field, instead I would inject here protocol version to factory construct from configuration/cli parameters and create an instance of factory that would be used multiple times. What do you think?

**Comment by findepi** [comment]
@AlekseiS @kokosing is there a documentation how protocol v2 differs from v1?
(i vaguely remember seeing something like that but I wouldn't assume it was completed or it matches implementation in 100%)

**Comment by sopel39** [comment]
@AlekseiS I am also curious is there a plan to add support for very large data (e.g: BLOBs), e.g: tens of MBs? Would be super useful if we wanted to store compiled methods for execution during query (@dain ?)

I know that worker to worker has page size limitation right now, but I would guess client protocol is another story.

**Comment by AlekseiS** [comment]
Rebased

**Comment by AlekseiS** [comment]
@sopel39  Do you want to support large data as part of protocol control information, like properties, prepared statements, etc? Or do you mean downloading large amounts of data by the client?

**Comment by AlekseiS** [comment]
Updated PR description to include a wiki link to protocol description and key differences with v1. Also, mention limitations of the current implementations.

File: presto-benchmark-driver/src/main/java/com/facebook/presto/benchmark/driver/BenchmarkQueryRunner.java [hunk 5]
```diff
@@ -151,18 +151,51 @@ public BenchmarkQueryResult execute(Suite suite, ClientSession session, Benchmar
         failures = 0;
         while (true) {
             // start query
-            StatementClient client = newStatementClient(okHttpClient, session, "show schemas");
+            try (StatementClient client = newStatementClient(okHttpClient, session, "show schemas")) {
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-benchmark-driver/src/main/java/com/facebook/presto/benchmark/driver/BenchmarkQueryRunner.java]: I see you already have a viable approach in your PR. I'll rebase on it once it's merged.

File: presto-benchmark-driver/src/main/java/com/facebook/presto/benchmark/driver/BenchmarkQueryRunner.java [hunk 6]
```diff
@@ -151,18 +151,51 @@ public BenchmarkQueryResult execute(Suite suite, ClientSession session, Benchmar
...
+                // read query output
+                ImmutableList.Builder<String> schemas = ImmutableList.builder();
+                while (client.isValid() && client.advance()) {
+                    // we do not process the output
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-benchmark-driver/src/main/java/com/facebook/presto/benchmark/driver/BenchmarkQueryRunner.java]: Probably not. I can remove it, but it's not related to my changes.

File: presto-cli/src/main/java/com/facebook/presto/cli/QueryRunner.java [hunk 7]
```diff
@@ -68,7 +68,7 @@ public QueryRunner(
 
         OkHttpClient.Builder builder = new OkHttpClient.Builder();
 
-        setupTimeouts(builder, 5, SECONDS);
+        setupTimeouts(builder, 30, SECONDS);
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-cli/src/main/java/com/facebook/presto/cli/QueryRunner.java]: Yes, it's going to affect both. Commit message just gives v2 as an explanation why. I think it shouldn't be a problem to increase it for both. However, if someone thinks otherwise I can configure different timeouts based on protocol version.

File: presto-client/src/main/java/com/facebook/presto/client/QueryActions.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    @JsonProperty
+    public Boolean isClearTransactionId()
+    {
+        return clearTransactionId;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-client/src/main/java/com/facebook/presto/client/QueryActions.java]: Done

File: presto-client/src/main/java/com/facebook/presto/client/QueryActions.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+                .add("clearSessionProperties", clearSessionProperties)
+                .add("addedPreparedStatements", addedPreparedStatements)
+                .add("deallocatedPreparedStatements", deallocatedPreparedStatements)
+                .add("startedTransactionId", startedTransactionId)
+                .add("clearTransactionId", clearTransactionId)
+                .toString();
+    }
+
+    @Nullable
+    public static QueryActions createIfNecessary(
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-client/src/main/java/com/facebook/presto/client/QueryActions.java]: So, that it can directly be passed as a field value in `QueryResults`. Also, it seems to be more inline with the current style of using nullable fields for classes used as json structures.

File: presto-client/src/main/java/com/facebook/presto/client/StatementClientFactory.java (hunk 4 above)
Comment (AlekseiS) [reply presto-client/src/main/java/com/facebook/presto/client/StatementClientFactory.java]: Both v1 and v2 might be used by different statement clients in the same process, for example unit tests use it. As a result, it cannot be a singleton initialized at start up. What class/component will then own an instance of the version-specific factory?

File: presto-client/src/main/java/com/facebook/presto/client/StatementClientFactory.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.client;
+
+import okhttp3.OkHttpClient;
+
+public final class StatementClientFactory
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-client/src/main/java/com/facebook/presto/client/StatementClientFactory.java]: I need to think about it. One thing I noticed is that CLI creates a new http client for every query by copying properties from existing one.  See `QueryRunner.startInternalQuery()`, line 118. @electrum , is it intentional?
If so then factory won't be able to own `httpClient`.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryExecutor.java [hunk 11]
```diff
@@ -42,7 +43,7 @@ public QueryExecutor(OkHttpClient httpClient)
 
     public StatementClient startQuery(ClientSession session, String query)
     {
-        return new StatementClient(httpClient, session, query);
+        return newStatementClient(httpClient, session, query);
     }
 
     public ServerInfo getServerInfo(URI server)
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryExecutor.java]: I think you're suggesting wider changes to how clients are managed across CLI, jdbc, etc. It might be possible, but I'd rather do it as a different PR.

File: presto-main/src/main/java/com/facebook/presto/server/SessionContext.java [hunk 12]
```diff
@@ -16,38 +16,55 @@
...
     Identity getIdentity();
 
+    @Nullable
     String getCatalog();
 
+    @Nullable
     String getSchema();
 
+    @Nullable
     String getSource();
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/server/SessionContext.java]: In general, yes, but somehow it was done this way. I added annotations to at least make it clear that some of these fields are nullable.

File: presto-main/src/main/java/com/facebook/presto/server/SessionContext.java [hunk 13]
```diff
@@ -16,38 +16,55 @@
...
 
+    @NotNull
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/server/SessionContext.java]: Removed.

File: presto-main/src/main/java/com/facebook/presto/server/TaskResource.java [hunk 14]
```diff
@@ -241,19 +248,7 @@ public void getResults(@PathParam("taskId") TaskId taskId,
             @Suspended AsyncResponse asyncResponse)
             throws InterruptedException
     {
-        requireNonNull(taskId, "taskId is null");
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/server/TaskResource.java]: Good idea. I moved method extraction in this class to a separate commit.

File: presto-main/src/main/java/com/facebook/presto/server/TaskResource.java [hunk 15]
```diff
@@ -326,4 +345,98 @@ private static Duration randomizeWaitTime(Duration waitTime)
...
+        private final TaskResource taskResource;
+
+        @Inject
+        public TaskDownloadResource(TaskResource taskResource)
+        {
+            this.taskResource = requireNonNull(taskResource, "taskResource is null");
+        }
+
+        @GET
+        @Path("{taskId}/results/{bufferId}/{token}")
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/server/TaskResource.java]: It's new in this class, because the logic is different: convert to json vs convert to pages. It uses the same approach with `RowIterable` as `StatementClient`.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java (hunk 1 above)
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java]: Well, createV1 and createV2 need different parameters. Public methods of `Query` abstract class is pretty much the interface. In any case, let me think a bit more about it.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/StatementResourceV2.java (hunk 2 above)
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/StatementResourceV2.java]: I didn't rename it mainly to make it easier to rebase and modify this PR. I suggest renaming it once this PR is merged.

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 16]
```diff
@@ -432,4 +427,50 @@ private static void closeUnchecked(AutoCloseable closeable)
...
+        private SqlParserOptions parserOptions = DEFAULT_SQL_PARSER_OPTIONS;
+        private String environment = ENVIRONMENT;
+
+        private Builder(Session defaultSession, int workersCount)
+        {
+            this.defaultSession = defaultSession;
+            this.workerCount = workersCount;
+        }
+
+        public Builder setExtraProperties(Map<String, String> extraProperties)
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: I checked the usage of `DistributedQueryRunner` and there seems to be only one case (the one you linked) where a single property is set for both extra and coordinator properties. However, it's much more common in `TpchQueryRunner`, so I added these methods there.

**Comment by AlekseiS** [comment]
@kokosing thanks for reviewing the changes! I made several changes and commented on the rest.

**Comment by kokosing** [comment]
Can you please rebase

**Comment by AlekseiS** [comment]
Rebased again.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/QueryV1.java [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static com.facebook.presto.SystemSessionProperties.isExchangeCompressionEnabled;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.util.Objects.requireNonNull;
+
+class QueryV1
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/QueryV1.java]: I think you could extract this class in separate commit.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java [hunk 18]
```diff
@@ -78,6 +83,8 @@
     private final AtomicReference<TaskHolder> taskHolderReference = new AtomicReference<>(new TaskHolder());
     private final AtomicBoolean needsPlan = new AtomicBoolean(true);
 
+    private final AtomicReference<Optional<TaskClientOutputContext>> clientOutputContext = new AtomicReference<>(Optional.empty());
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java]: Getter is not synchronized, so the field needs to be marked as "volatile" for an update to be seen by another thread. Throughout Presto codebase I see atomic primitives are commonly used in such cases instead of volatile.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java [hunk 19]
```diff
@@ -333,6 +345,15 @@ public TaskInfo updateTask(Session session, Optional<PlanFragment> fragment, Lis
                 taskExecution = taskHolder.getTaskExecution();
                 if (taskExecution == null) {
                     checkState(fragment.isPresent(), "fragment must be present");
+                    // this block must happen before creating task execution to prevent a race condition when execution finished,
+                    // but types are not ready. Types are accessed by a task resource in order to convert pages to a client output
+                    if (fragment.get().getRoot() instanceof OutputNode) {
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java]: My understanding is that this whole synchronized block happens only once per task because it creates task execution. However, I added a check as you suggested.

File: presto-client/src/main/java/com/facebook/presto/client/QueryActions.java [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+import javax.annotation.concurrent.Immutable;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+
+@Immutable
+public class QueryActions
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-client/src/main/java/com/facebook/presto/client/QueryActions.java]: done

File: presto-main/src/main/java/com/facebook/presto/server/TaskClientOutputContext.java [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+import com.facebook.presto.execution.buffer.PagesSerde;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class TaskClientOutputContext
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/server/TaskClientOutputContext.java]: Done

File: presto-main/src/main/java/com/facebook/presto/server/protocol/QueryV1.java (hunk 17 above)
Comment (AlekseiS) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/QueryV1.java]: Doing it in a separate commit won't be easy as the structure of a `Query` is changed and the changes are closely related to introduction of v2 protocol and changes in `StatementResource`.

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 22]
```diff
@@ -412,6 +402,11 @@ public final void close()
         }
     }
 
+    public static Builder builder(Session defaultSession, int workersCount)
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: Done

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 23]
```diff
@@ -91,10 +92,10 @@ public DistributedQueryRunner(Session defaultSession, int workersCount)
     public DistributedQueryRunner(Session defaultSession, int workersCount, Map<String, String> extraProperties)
```
[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: Done

**Comment by AlekseiS** [comment]
Rebased and addressed some review comments. Remaining items from review:
* deduplicate parsing logic in `QueryRequestSessionContext`
* think about extending `DistributedQueryRunner.Builder` in `TpchQueryRunner`
* add v2 protocol to product tests

@kokosing , regarding migration, as the first step both protocols can be enabled at the same time. The change allows them to co-exist. During this time some clients/clusters can be switched to using it. Once there's enough confidence, "v2" can be set as the default protocol version in command line and jdbc (once implemented). Later "v1" can be disabled.

BTW, I'm ok merging this PR in parts if you feel more comfortable with it.

**Comment by kokosing** [comment]
> BTW, I'm ok merging this PR in parts if you feel more comfortable with it.

v2 is behind the feature flag, so it should be quite safe to merge it as it. We would need just make sure that there is no regression introduced to v1. Anyway, I think it is desired to have someone from FB to take a look at it as well, especially someone well familiar with this code.

**Comment by AlekseiS** [comment]
Rebased on 0.193.

**Comment by AlekseiS** [comment]
Rebased again to resolve conflicts.

**Comment by AlekseiS** [comment]
Deduplicated parsing logic in `QueryRequestSessionContext`.

**Comment by findepi** [comment]
> v2 is behind the feature flag

Is it possible to run v1 and v2 side-by-side? That would make the change easier to integrate.

**Comment by AlekseiS** [comment]
@findepi , when enabled, v2 will run side by side with v1 on the server.

**Comment by kokosing** [comment]
> when enabled, v2 will run side by side with v1 on the server.

But why it is not a default behaviour ? Does v2 affects somehow queries which are using v1?

**Comment by AlekseiS** [comment]
> why it is not a default behaviour

@kokosing , I think the main reason is that the protocol is not finished yet, so there might be more changes to it. Other than that, there's no problem enabling it side-by-side with v1.

**Comment by AlekseiS** [comment]
Rebased.

**Comment by AlekseiS** [comment]
Add a new commit to send dataUris in a cumulative way as they become available.

File: presto-client/src/main/java/com/facebook/presto/client/ProtocolVersion.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+
+import static java.util.Objects.requireNonNull;
+
+public enum ProtocolVersion
+{
+    V1, V2;
+
+    public static ProtocolVersion fromString(String value)
+    {
+        requireNonNull(value, "value is null");
```
Comment (electrum) [inline presto-client/src/main/java/com/facebook/presto/client/ProtocolVersion.java]: This can be
```java
return valueOf(value.toUpperCase(ENGLISH));

Comment (AlekseiS) [reply presto-client/src/main/java/com/facebook/presto/client/ProtocolVersion.java]: Done

**Comment by AlekseiS** [comment]
Rebased on (almost) 0.195.
Squashed last commit changing how dataUris are sent, updated wiki.

**Comment by AlekseiS** [comment]
Extend `DistributedQueryRunner.Builder` in `TpchQueryRunnerBuilder`.
@kokosing see if the corresponding two commits look better now.

**Comment by kokosing** [comment]
@AlekseiS Thank you.

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueues.java [hunk 25]
```diff
@@ -284,7 +284,7 @@ public void testClientTagsBasedSelection()
     public void testQueryTypeBasedSelection()
             throws Exception
     {
-        try (DistributedQueryRunner queryRunner = TpchQueryRunner.createQueryRunner(ImmutableMap.of(), ImmutableMap.of("experimental.resource-groups-enabled", "true"))) {
+        try (DistributedQueryRunner queryRunner = TpchQueryRunnerBuilder.builder().setSingleCoordinatorProperty("experimental.resource-groups-enabled", "true").build()) {
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueues.java]: Wrap here

File: presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java [hunk 26]
```diff
@@ -1,108 +0,0 @@
...
-import com.facebook.presto.tpch.TpchPlugin;
-import com.google.common.collect.ImmutableMap;
-import io.airlift.log.Logger;
-import io.airlift.log.Logging;
-
-import java.util.Map;
-
-import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
-
-public final class TpchQueryRunner
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java]: Let's keep the class around with just the main method

File: presto-main/src/main/java/com/facebook/presto/server/SessionContext.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (electrum) [reply presto-main/src/main/java/com/facebook/presto/server/SessionContext.java]: Yes, it would be nice to convert these to optionals, but that's a bigger project.
